File: kant_dao/src/kant_dao/ros2_dao/ros2_pddl_object_dao.cpp

#include "kant_msgs/msg/update_knowledge.hpp"

#include "kant_dao/ros2_dao/ros2_pddl_object_dao.hpp"

using namespace kant::dao::ros2_dao;

Ros2PddlObjectDao::Ros2PddlObjectDao(simple_node::Node *node) {

  this->dto_msg_parser =
      std::make_unique<kant::knowledge_base::parser::DtoMsgParser>();
  this->msg_dto_parser =
      std::make_unique<kant::knowledge_base::parser::MsgDtoParser>();

  // srv clients
  this->get_client =
      node->create_client<kant_msgs::srv::GetPddlObject>("get_objects");

  this->update_client =
      node->create_client<kant_msgs::srv::UpdatePddlObject>(
          "update_object");

  this->delete_all_client =
      node->create_client<std_srvs::srv::Empty>("delete_all_objects");
}

std::vector<std::shared_ptr<kant::dto::PddlObjectDto>>
Ros2PddlObjectDao::ros2_get(std::string object_name) {

  auto request =
      std::make_shared<kant_msgs::srv::GetPddlObject::Request>();

  request->object_name = object_name;

  this->get_client->wait_for_service();
  auto future = this->get_client->async_send_request(request);
  future.wait();
  auto result = future.get();

  std::vector<std::shared_ptr<kant::dto::PddlObjectDto>> pddl_object_dto_list;

  for (const auto &pddl_object_msg : result->pddl_objects) {
    auto pddl_object_dto =
        this->msg_dto_parser->object_msg_to_dto(pddl_object_msg);
    pddl_object_dto_list.push_back(pddl_object_dto);
  }

  return pddl_object_dto_list;
}

bool Ros2PddlObjectDao::ros2_update(
    std::shared_ptr<kant::dto::PddlObjectDto> pddl_object_dto,
    int update_type) {

  auto request =
      std::make_shared<kant_msgs::srv::UpdatePddlObject::Request>();

  request->pddl_object =
      this->dto_msg_parser->object_dto_to_msg(pddl_object_dto);
  request->update_konwledge.update_type = update_type;

  this->update_client->wait_for_service();
  auto future = this->update_client->async_send_request(request);
  future.wait();
  auto result = future.get();

  return result->success;
}

void Ros2PddlObjectDao::ros2_delete_all() {

  auto request = std::make_shared<std_srvs::srv::Empty::Request>();
  this->delete_all_client->wait_for_service();
  auto future = this->delete_all_client->async_send_request(request);
  future.wait();
}

std::shared_ptr<kant::dto::PddlObjectDto>
Ros2PddlObjectDao::get(std::string object_name) {

  auto pddl_object_dto_list = this->ros2_get(object_name);

  if (pddl_object_dto_list.size() == 1) {
    return pddl_object_dto_list.at(0);
  }

  return nullptr;
}

std::vector<std::shared_ptr<kant::dto::PddlObjectDto>>
Ros2PddlObjectDao::get_all() {

  auto pddl_object_dto_list = this->ros2_get("");
  return pddl_object_dto_list;
}

bool Ros2PddlObjectDao::save(
    std::shared_ptr<kant::dto::PddlObjectDto> pddl_object_dto) {

  if (this->get(pddl_object_dto->get_name()) == nullptr) {
    return this->ros2_update(pddl_object_dto,
                             kant_msgs::msg::UpdateKnowledge::SAVE);
  }

  return false;
}

bool Ros2PddlObjectDao::update(
    std::shared_ptr<kant::dto::PddlObjectDto> pddl_object_dto) {
  if (this->get(pddl_object_dto->get_name()) != nullptr) {
    return this->ros2_update(pddl_object_dto,
                             kant_msgs::msg::UpdateKnowledge::SAVE);
  }

  return false;
}

bool Ros2PddlObjectDao::save_update(
    std::shared_ptr<kant::dto::PddlObjectDto> pddl_object_dto) {
  bool succ;
  if (this->get(pddl_object_dto->get_name()) == nullptr) {
    succ = this->save(pddl_object_dto);
  } else {
    succ = this->update(pddl_object_dto);
  }

  return succ;
}

bool Ros2PddlObjectDao::delete_one(
    std::shared_ptr<kant::dto::PddlObjectDto> pddl_object_dto) {
  return this->ros2_update(pddl_object_dto,
                           kant_msgs::msg::UpdateKnowledge::DELETE);
}

bool Ros2PddlObjectDao::delete_all() {

  this->ros2_delete_all();
  return true;
}

File: kant_dao/kant_dao/ros2_dao/ros2_pddl_object_dao.py

""" Ros2 Pddl Object Dao """

from typing import List
from simple_node import Node

from kant_dao.dao_interface import PddlObjectDao
from kant_dto import PddlObjectDto

from kant_msgs.srv import (
    UpdatePddlObject,
    GetPddlObject
)
from kant_msgs.msg import UpdateKnowledge
from std_srvs.srv import Empty

from kant_knowledge_base.parser import (
    DtoMsgParser,
    MsgDtoParser
)


class Ros2PddlObjectDao(PddlObjectDao):
    """ Ros2 Pddl Object Dao Class """

    def __init__(self, node: Node):

        PddlObjectDao.__init__(self)

        self.node = node

        # parsers
        self.dto_msg_parser = DtoMsgParser()
        self.msg_dto_parser = MsgDtoParser()

        # srv clients
        self._get_client = self.node.create_client(
            GetPddlObject, "get_objects")

        self._update_client = self.node.create_client(
            UpdatePddlObject, "update_object")

        self._delete_all_client = self.node.create_client(
            Empty, "delete_all_objects")

    def _ros2_get(self, object_name: str = "") -> List[PddlObjectDto]:
        """ ros2_get method

        Args:
            object_name (str): object name

        Returns:
            List[PddlObjectDto]: list of PddlObjectDto
        """

        req = GetPddlObject.Request()

        req.object_name = object_name

        self._get_client.wait_for_service()
        result = self._get_client.call(req)

        pddl_object_dto_list = []

        for pddl_object_msg in result.pddl_objects:
            pddl_object_dto = self.msg_dto_parser.object_msg_to_dto(
                pddl_object_msg)
            pddl_object_dto_list.append(pddl_object_dto)

        return pddl_object_dto_list

    def _ros2_update(self,
                     pddl_object_dto: PddlObjectDto,
                     update_type: int) -> bool:
        """ ros2_delete_all method

        Args:
            pddl_object_dto (PddlObjectDto): PddlObjectDto to update

        Returns:
            bool: succeed
        """

        req = UpdatePddlObject.Request()

        req.pddl_object = self.dto_msg_parser.object_dto_to_msg(
            pddl_object_dto)
        req.update_konwledge.update_type = update_type

        self._update_client.wait_for_service()
        result = self._update_client.call(req)

        return result.success

    def _ros2_delete_all(self):
        """ asyn delete_all method

        Returns:
            bool: succeed
        """

        req = Empty.Request()

        self._delete_all_client.wait_for_service()
        self._delete_all_client.call(req)

    def get(self, object_name: str) -> PddlObjectDto:
        """ get a PddlObjectDto with a given object name
            return None if there is no pddl with that object name

        Args:
            object_name (str): pddl object name

        Returns:
            PddlObjectDto: PddlObjectDto of the pddl object name
        """

        pddl_object_dto_list = self._ros2_get(object_name)

        if len(pddl_object_dto_list) == 1:
            return pddl_object_dto_list[0]

        return None

    def get_all(self) -> List[PddlObjectDto]:
        """ get all PddlObjectDto

        Returns:
            List[PddlObjectDto]: list of all PddlObjectDto
        """

        pddl_object_dto_list = self._ros2_get()

        return pddl_object_dto_list

    def _save(self, pddl_object_dto: PddlObjectDto) -> bool:
        """ save a PddlObjectDto
            if the PddlObjectDto is already saved return False, else return True

        Args:
            pddl_object_dto (PddlObjectDto): PddlObjectDto to save

        Returns:
            bool: succeed
        """

        if not self.get(pddl_object_dto.get_name()):
            succ = self._ros2_update(
                pddl_object_dto, UpdateKnowledge.SAVE)
            return succ

        return False

    def _update(self, pddl_object_dto: PddlObjectDto) -> bool:
        """ update a PddlObjectDto
            if the PddlObjectDto is not saved return False, else return True

        Args:
            pddl_object_dto (PddlObjectDto): PddlObjectDto to update

        Returns:
            bool: succeed
        """

        if self.get(pddl_object_dto.get_name()):
            succ = self._ros2_update(
                pddl_object_dto, UpdateKnowledge.SAVE)
            return succ

        return False

    def save(self, pddl_object_dto: PddlObjectDto) -> bool:
        """ save or update a PddlObjectDto
            if the PddlObjectDto is not saved it will be saved, else it will be updated

        Args:
            pddl_object_dto (PddlObjectDto): PddlObjectDto to save or update

        Returns:
            bool: succeed
        """

        if not self.get(pddl_object_dto.get_name()):
            succ = self._save(pddl_object_dto)
        else:
            succ = self._update(pddl_object_dto)

        return succ

    def delete(self, pddl_object_dto: PddlObjectDto) -> bool:
        """ delete a PddlObjectDto
            if the PddlObjectDto is not saved return False, else return True

        Args:
            pddl_object_dto (PddlObjectDto): PddlObjectDto to delete

        Returns:
            bool: succeed
        """

        succ = self._ros2_update(
            pddl_object_dto, UpdateKnowledge.DELETE)
        return succ

    def delete_all(self) -> bool:
        """ delete all pddl objects

        Returns:
            bool: succeed
        """

        self._ros2_delete_all()

        return True
